File: app/src/main/java/deltazero/amarok/ui/MainActivity.java
package deltazero.amarok.ui;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.progressindicator.CircularProgressIndicator;
import com.hjq.permissions.XXPermissions;

import deltazero.amarok.AmarokActivity;
import deltazero.amarok.Hider;
import deltazero.amarok.PrefMgr;
import deltazero.amarok.R;
import deltazero.amarok.apphider.NoneAppHider;
import deltazero.amarok.filehider.NoneFileHider;
import deltazero.amarok.ui.settings.SettingsActivity;
import deltazero.amarok.ui.settings.SwitchAppHiderActivity;
import deltazero.amarok.ui.settings.SwitchFileHiderActivity;
import deltazero.amarok.utils.PermissionUtil;
import nl.dionsegijn.konfetti.xml.KonfettiView;

public class MainActivity extends AmarokActivity {

    public final static String TAG = "Main";
    private ImageView ivStatusImg;
    private TextView tvStatusInfo, tvStatus, tvMoto;
    private MaterialButton btChangeStatus, btSetHideFiles, btSetHideApps;
    private CircularProgressIndicator piProcessStatus;
    private KonfettiView konfettiView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        // Setup activity
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Binding views
        ivStatusImg = findViewById(R.id.main_iv_status);
        tvStatus = findViewById(R.id.main_tv_status);
        tvStatusInfo = findViewById(R.id.main_tv_statusinfo);
        tvMoto = findViewById(R.id.main_tv_moto);
        btChangeStatus = findViewById(R.id.main_bt_change_status);
        btSetHideApps = findViewById(R.id.main_bt_set_hide_apps);
        btSetHideFiles = findViewById(R.id.main_bt_set_hide_files);
        piProcessStatus = findViewById(R.id.main_pi_process_status);
        konfettiView = findViewById(R.id.main_konfetti_view);

        // Init UI
        refreshUi(Hider.getState());

        // Setup observer
        Hider.state.observe(this, this::refreshUi);

        // Show welcome dialog
        if (PrefMgr.getShowWelcome()) {
            new MaterialAlertDialogBuilder(this)
                    .setTitle(R.string.welcome_title)
                    .setMessage(R.string.welcome_msg)
                    .setPositiveButton(R.string.ok, (dialog, which)
                            -> PermissionUtil.requestStoragePermission(this))
                    .setNegativeButton(R.string.view_github_repo, (dialog, which) -> {
                        startActivity(new Intent(Intent.ACTION_VIEW,
                                Uri.parse("https://github.com/deltazefiro/Amarok-Hider")));
                        PermissionUtil.requestStoragePermission(this);
                    })
                    .setOnCancelListener(dialog -> PermissionUtil.requestStoragePermission(this))
                    .show();
            PrefMgr.setShowWelcome(false);
        } else {
            PermissionUtil.requestStoragePermission(this);
        }

        // Check Hiders availability
        PrefMgr.getAppHider(this).tryToActivate((appHiderClass, succeed, msg) -> {
            if (succeed) return;
            Hider.showNoHiderDialog(this, msg);
        });

        PrefMgr.getFileHider(this).tryToActive((fileHiderClass, succeed, msg) -> {
            if (succeed) return;
            PrefMgr.setFileHiderMode(NoneFileHider.class);
            new MaterialAlertDialogBuilder(this)
                    .setTitle(R.string.filehider_not_ava_title)
                    .setMessage(msg)
                    .setPositiveButton(R.string.switch_file_hider, (dialog, which)
                            -> startActivity(new Intent(this, SwitchFileHiderActivity.class)))
                    .setNegativeButton(getString(R.string.ok), null)
                    .show();
        });
    }

    public void changeStatus(View view) {
        if (Hider.getState() == Hider.State.HIDDEN) Hider.unhide(this);
        else Hider.hide(this);
    }

    public void setHideApps(View view) {

        if (Hider.getState() == Hider.State.HIDDEN) {
            Toast.makeText(this, R.string.setting_not_ava_when_hidden, Toast.LENGTH_SHORT).show();
            return;
        }

        if (PrefMgr.getAppHider(this) instanceof NoneAppHider) {
            new MaterialAlertDialogBuilder(this)
                    .setTitle(R.string.apphider_not_activated_title)
                    .setMessage(R.string.apphider_not_activated_msg)
                    .setPositiveButton(R.string.switch_app_hider, (dialog, which)
                            -> startActivity(new Intent(this, SwitchAppHiderActivity.class)))
                    .setNegativeButton(getString(R.string.cancel), null)
                    .show();
            return;
        }

        startActivity(new Intent(this, SetHideAppActivity.class));

    }

    public void showMoreSettings(View view) {

        startActivity(new Intent(this, SettingsActivity.class));

    }

    public void setHideFile(View view) {

        if (!XXPermissions.isGranted(this, com.hjq.permissions.Permission.MANAGE_EXTERNAL_STORAGE)) {
            Toast.makeText(this, R.string.storage_permission_denied, Toast.LENGTH_LONG).show();
            return;
        }

        if (Hider.getState() == Hider.State.HIDDEN) {
            Toast.makeText(this, R.string.setting_not_ava_when_hidden, Toast.LENGTH_SHORT).show();
            return;
        }

        startActivity(new Intent(this, SetHideFilesActivity.class));
    }

    public void refreshUi(Hider.State state) {
        tvMoto.setText(R.string.moto);
        switch (state) {
            case HIDDEN -> {
                // Not Processing
                piProcessStatus.hide();
                btChangeStatus.setEnabled(true);
                // Hidden
                ivStatusImg.setImageResource(R.drawable.img_status_hidden);
                ivStatusImg.setImageTintList(getColorStateList(com.google.android.material.R.color.material_on_background_emphasis_high_type));
                btChangeStatus.setText(R.string.unhide);
                btChangeStatus.setIconResource(R.drawable.ic_wolf);
                btSetHideFiles.setEnabled(false);
                btSetHideApps.setEnabled(false);
                tvStatus.setText(getText(R.string.hidden_status));
                tvStatusInfo.setText(getText(R.string.hidden_moto));
            }
            case VISIBLE -> {
                // Not Processing
                piProcessStatus.hide();
                btChangeStatus.setEnabled(true);
                // Visible
                ivStatusImg.setImageResource(R.drawable.img_status_visible);
                ivStatusImg.setImageTintList(null);
                btChangeStatus.setText(R.string.hide);
                btChangeStatus.setIconResource(R.drawable.ic_paw);
                btSetHideFiles.setEnabled(true);
                btSetHideApps.setEnabled(true);
                tvStatus.setText(getText(R.string.visible_status));
                tvStatusInfo.setText(getText(R.string.visible_moto));
            }
            case PROCESSING -> {
                // Processing
                piProcessStatus.show();
                btChangeStatus.setEnabled(false);
            }
        }
    }

    @Override
    protected void onResume() {
        refreshUi(Hider.getState());
        super.onResume();
    }
}

File: app/src/main/java/deltazero/amarok/ui/MainActivity.kt
package deltazero.amarok.ui

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import androidx.activity.compose.setContent
import androidx.activity.viewModels
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.WindowInsetsSides
import androidx.compose.foundation.layout.only
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.systemBars
import androidx.compose.material3.Scaffold
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.lifecycle.lifecycleScope
import androidx.navigation.NavHostController
import androidx.navigation.compose.NavHost
import androidx.navigation.compose.composable
import androidx.navigation.compose.currentBackStackEntryAsState
import androidx.navigation.compose.rememberNavController
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.hjq.permissions.OnPermissionCallback
import com.skydoves.colorpickerview.ColorPickerDialog
import com.skydoves.colorpickerview.listeners.ColorEnvelopeListener
import deltazero.amarok.AmarokActivity
import deltazero.amarok.R
import deltazero.amarok.core.Hider
import deltazero.amarok.ui.settings.SettingsScreen
import deltazero.amarok.ui.settings.SettingsViewModel
import deltazero.amarok.ui.theme.AmarokTheme
import deltazero.amarok.utils.PermissionUtil
import deltazero.amarok.utils.UpdateUtil
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class MainActivity : AmarokActivity() {

  private val settingsViewModel: SettingsViewModel by viewModels()
  private lateinit var navController: NavHostController

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    setContent {
      AmarokTheme {
        navController = rememberNavController()
        val backStackEntry by navController.currentBackStackEntryAsState()
        val currentRoute = backStackEntry?.destination?.route
        val showBottomBar =
          currentRoute in AmarokRoute.tabRoutes || currentRoute == AmarokRoutes.APP_PICKER

        Scaffold(
          bottomBar = { if (showBottomBar) AmarokNavigationBar(navController) },
          // Let each screen's top app bar own the status-bar area so it draws
          // edge-to-edge behind it; only the bottom/horizontal insets are consumed here.
          contentWindowInsets =
            WindowInsets.systemBars.only(WindowInsetsSides.Horizontal + WindowInsetsSides.Bottom),
        ) { padding ->
          NavHost(
            navController = navController,
            startDestination = AmarokRoute.DASHBOARD.route,
            modifier = Modifier.padding(padding),
            enterTransition = { fadeIn(tween(500)) },
            exitTransition = { fadeOut(tween(500)) },
          ) {
            composable(AmarokRoute.DASHBOARD.route) {
              DashboardScreen(onChangeStatus = { changeStatus() })
            }
            composable(AmarokRoute.APPS.route) {
              AppsScreen(onOpenEditor = { navController.navigate(AmarokRoutes.APP_PICKER) })
            }
            composable(AmarokRoute.FILES.route) { FilesScreen() }
            composable(AmarokRoutes.APP_PICKER) {
              AppPickerScreen(onBack = { navController.popBackStack() })
            }
            composable(AmarokRoute.SETTINGS.route) {
              SettingsScreen(
                onSetPassword = { callback ->
                  SetPasswordFragment()
                    .setCallback { password -> callback(password) }
                    .show(supportFragmentManager, null)
                },
                onShowCountdownConfirm = { onConfirm, onCancel ->
                  CountdownConfirmDialog.Builder(this@MainActivity)
                    .setTitle(R.string.hide_amarok_icon_dialog_title)
                    .setMessage(R.string.hide_amarok_icon_dialog_message)
                    .setCountdownSeconds(10)
                    .setOnConfirmAction(onConfirm)
                    .setOnCancelAction(onCancel)
                    .show()
                },
                onRequestNotificationPermission = { onGranted, onDenied ->
                  PermissionUtil.requestNotificationPermission(
                    this@MainActivity,
                    object : OnPermissionCallback {
                      override fun onGranted(permissions: MutableList<String>, all: Boolean) {
                        onGranted()
                      }

                      override fun onDenied(permissions: MutableList<String>, never: Boolean) {
                        onDenied()
                      }
                    },
                  )
                },
                onRequestSystemAlertPermission = { onGranted, onDenied ->
                  PermissionUtil.requestSystemAlertPermission(
                    this@MainActivity,
                    object : OnPermissionCallback {
                      override fun onGranted(permissions: MutableList<String>, all: Boolean) {
                        onGranted()
                      }

                      override fun onDenied(permissions: MutableList<String>, never: Boolean) {
                        onDenied()
                      }
                    },
                  )
                },
                onShowColorPicker = {
                  val builder =
                    ColorPickerDialog.Builder(this@MainActivity)
                      .setTitle(R.string.panic_button_color)
                      .setPreferenceName("PanicButtonColorPicker")
                      .setPositiveButton(
                        getString(android.R.string.ok),
                        ColorEnvelopeListener { envelope, _ ->
                          lifecycleScope.launch(Dispatchers.IO) {
                            settingsRepo.setPanicButtonColor(envelope.color)
                          }
                        },
                      )
                      .setNegativeButton(getString(android.R.string.cancel)) { dialog, _ ->
                        dialog.dismiss()
                      }
                      .attachAlphaSlideBar(true)
                      .attachBrightnessSlideBar(true)
                      .setBottomSpace(12)
                  builder.colorPickerView.setInitialColor(
                    settingsRepo.settings.value.panicButtonColor
                  )
                  builder.show()
                },
                viewModel = settingsViewModel,
              )
            }
          }
        }
      }
    }

    // Show welcome dialog
    if (settingsRepo.settings.value.showWelcome) {
      MaterialAlertDialogBuilder(this)
        .setTitle(R.string.welcome_title)
        .setMessage(R.string.welcome_msg)
        .setPositiveButton(R.string.ok) { _, _ -> PermissionUtil.requestStoragePermission(this) }
        .setNegativeButton(R.string.view_github_repo) { _, _ ->
          startActivity(
            Intent(Intent.ACTION_VIEW, Uri.parse("https://github.com/deltazefiro/Amarok-Hider"))
          )
          PermissionUtil.requestStoragePermission(this)
        }
        .setOnCancelListener { PermissionUtil.requestStoragePermission(this) }
        .show()
      lifecycleScope.launch(Dispatchers.IO) { settingsRepo.setShowWelcome(false) }
    } else {
      PermissionUtil.requestStoragePermission(this)
    }
    if (settingsRepo.settings.value.autoUpdate) {
      UpdateUtil.checkAndNotify(this, true)
    }
  }

  private fun changeStatus() {
    val appErr = Hider.appHiderError.value
    if (appErr != 0) {
      showNoHiderDialog(appErr)
      return
    }
    val fileErr = Hider.fileHiderError.value
    if (fileErr != 0) {
      showNoHiderDialog(fileErr)
      return
    }
    if (Hider.getState() == Hider.State.HIDDEN) {
      Hider.processAll(this, Hider.Action.UNHIDE)
    } else {
      Hider.processAll(this, Hider.Action.HIDE)
    }
  }

  private fun navigateToSettings() {
    navController.navigate(AmarokRoute.SETTINGS.route) {
      popUpTo(navController.graph.startDestinationId) { saveState = true }
      launchSingleTop = true
      restoreState = true
    }
  }

  private fun showNoHiderDialog(msgResID: Int) {
    MaterialAlertDialogBuilder(this)
      .setTitle(R.string.apphider_not_ava_title)
      .setMessage(msgResID)
      .setPositiveButton(R.string.more_settings) { _, _ -> navigateToSettings() }
      .setNegativeButton(getString(R.string.ok), null)
      .show()
  }
}
